client/packages/system/src/Item/api/hooks/bundledItem/useUpsertBundledItem.ts

```diff
@@ -0,0 +1,74 @@
+import { useState } from 'react';
+import {
+  FnUtils,
+  isEmpty,
+  useMutation,
+  useTranslation,
+} from '@openmsupply-client/common';
+import {
+  BundledItemFragment,
+  ItemVariantFragment,
+} from '../../operations.generated';
+import { useItemApi, useItemGraphQL } from '../useItemApi';
+
+export type DraftBundle = {
+  itemId: string;
+  variantId: string;
+  ratio: number;
+};
+
+export function useUpsertBundledItem({
+  principalVariant,
+  bundle,
+}: {
+  principalVariant: ItemVariantFragment;
+  bundle: BundledItemFragment | null;
+}) {
+  const { api, storeId, queryClient } = useItemGraphQL();
+  const { keys } = useItemApi();
+  const t = useTranslation();
+
+  const [draft, setDraft] = useState<DraftBundle>({
+    itemId: bundle?.bundledItemVariant?.itemId ?? '',
+    variantId: bundle?.bundledItemVariant?.id ?? '',
+    ratio: bundle?.ratio ?? 1,
+  });
+
+  const { mutateAsync } = useMutation({
+    mutationFn: async () => {
+      const apiResult = await api.upsertBundledItem({
+        storeId,
+        input: {
+          id: bundle?.id ?? FnUtils.generateUUID(),
+          principalItemVariantId: principalVariant.id,
+          bundledItemVariantId: draft.variantId,
+          ratio: draft.ratio,
+        },
+      });
+      // will be empty if there's a generic error, such as permission denied
+      if (!isEmpty(apiResult)) {
+        const result = apiResult.centralServer.bundledItem.upsertBundledItem;
+        if (result.__typename === 'BundledItemNode') {
+          return result;
+        }
+      }
+      throw new Error(t('error.failed-to-save-bundled-item'));
+    },
+    onSuccess: () => {
+      queryClient.invalidateQueries(keys.detail(principalVariant.itemId));
+    },
+  });
+
+  return {
+    draft,
+    isComplete: getIsComplete(draft),
```

Comment:
I was a bit unsure what we meant by "complete" -- perhaps `canSubmit` would be more specific? Then again, I see this name has been used in a few places now, so probably not worth changing them all. See what you think.